cmd/snapd-generator/main.c

```diff
@@ -30,30 +31,168 @@
+	for (size_t i = 0; i < sizeof dirs/sizeof(char *); ++i) {
+		if (mkdir(dirs[i], 0755) != 0 && errno != EEXIST) {
+			fprintf(stderr, "cannot create %s directory: %m\n", dirs[i]);
+			return 1;
+		}
+	}
+
+	char target[PATH_MAX + 1] = { 0 };
+	char lnpath[PATH_MAX + 1] = { 0 };
+	sc_must_snprintf(target, sizeof target, "../%s.mount", unit_name);
+	sc_must_snprintf(lnpath, sizeof lnpath,
+			 SYSTEMD_GENERATORS_SYSINIT_DIR "/%s.mount", unit_name);
+	if (symlink(target, lnpath) != 0) {
+		fprintf(stderr, "cannot create symlink %s: %m\n", lnpath);
+		return 1;
+	}
+
+	return 0;
+}
 
-	sc_mountinfo_entry *root = find_root_mountinfo(mounts);
+static int ensure_kernel_drivers_mounts(const char *normal_dir)
+{
+	// Find active kernel name and revision by looking at what was
+	// mounted in /run/mnt/kernel by snap-bootstrap.
+
+	const char* kern_mnt_dir = "/run/mnt/kernel";
+	sc_mountinfo_entry *kern_minfo = find_dir_mountinfo(kern_mnt_dir);
+	if (!kern_minfo) {
+		// This is not UBuntu Core / hybrid, do nothing and do not fail
+		return 0;
+	}
+
+	// Mount source should be a snap
+	if (!sc_streq(kern_minfo->fs_type, "squashfs")) {
+		fprintf(stderr, "unexpected fs type (%s) for %s\n",
+			kern_minfo->fs_type, kern_mnt_dir);
+		return 1;
+	}
+
+	// We expect a loop device as source (/dev/loop<number>)
+	size_t src_len = strlen(kern_minfo->mount_source);
+	static const char dev_pref[] = "/dev/loop";
+	static const size_t dev_pref_len = sizeof dev_pref - 1;
+	if (src_len < dev_pref_len + 1 || strncmp(kern_minfo->mount_source,
+						  "/dev/loop", dev_pref_len) != 0) {
+		fprintf(stderr, "unexpected mount source %s for %s\n",
+			kern_minfo->mount_source, kern_mnt_dir);
+		return 1;
+	}
+
+	// Find out backing file
+	char fname[PATH_MAX + 1] = { 0 };
+	// We ignore the "/dev/" prefix
+	static const char dev_dir[] = "/dev/";
+	sc_must_snprintf(fname, sizeof fname,
+			 "/sys/block/%s/loop/backing_file",
+			 kern_minfo->mount_source + sizeof dev_dir - 1);
+	FILE *f SC_CLEANUP(sc_cleanup_file) = NULL;
+	f = fopen(fname, "r");
+	if (!f) {
+		fprintf(stderr, "cannot open %s: %m\n", fname);
+		return 1;
+	}
+	char snap_path[PATH_MAX + 1] = { 0 };
+	if (fgets(snap_path, sizeof snap_path, f) == NULL) {
+		fprintf(stderr, "while reading %s: %m\n", fname);
+		return 1;
+	}
+
+	// Now parse the snap path
+	size_t i;
+	for (i = strlen(snap_path); i > 0 && snap_path[--i] != '/';);
+	char *snap_fname = snap_path + i + 1;
+
+	// snap_fname is of form <name>_<rev>.snap
+	char *saveptr = NULL;
+	char *snap_name = strtok_r(snap_fname, "_", &saveptr);
+	if (snap_name == NULL) {
+		fprintf(stderr, "snap name not found in loop backing file\n");
+		return 1;
+	}
+	char *snap_rev = strtok_r(NULL, ".", &saveptr);
+	if (snap_rev == NULL) {
+		fprintf(stderr, "snap revision not found in loop backing file\n");
+		return 1;
+	}
+
+	int res;
+	char what[PATH_MAX + 1] = { 0 };
+	sc_must_snprintf(what, sizeof what,
+			 "/var/lib/snapd/kernel/%s/%s/lib/modules",
```

Comment:
Maybe we should have a `#define` for `/var/lib/snapd/kernel`. And maybe make it use `LIBEXECDIR`.

Reply:
Sorry not `LIBEXECDIR`. We need `/var` to use `localstatedir` from autoconf though.

Reply:
Just forget about messing with autoconf. I think we have a mess here. I will look into it one day. Just a define is fine.

Reply:
Done